src/pyclass.rs

```diff
@@ -30,122 +31,157 @@ pub trait PyClass:
     type Frozen: Frozen;
 }
 
-fn into_raw<T>(vec: Vec<T>) -> *mut c_void {
-    Box::into_raw(vec.into_boxed_slice()) as _
-}
-
 pub(crate) fn create_type_object<T>(py: Python<'_>) -> *mut ffi::PyTypeObject
 where
     T: PyClass,
 {
     match unsafe {
-        create_type_object_impl(
-            py,
-            T::DOC,
-            T::MODULE,
-            T::NAME,
-            T::BaseType::type_object_raw(py),
-            std::mem::size_of::<T::Layout>(),
-            tp_dealloc::<T>,
-            T::dict_offset(),
-            T::weaklist_offset(),
-            T::items_iter,
-            T::IS_BASETYPE,
-            T::IS_MAPPING,
-        )
+        PyTypeBuilder::default()
+            .with_type_doc(T::DOC)
+            .with_offsets(T::dict_offset(), T::weaklist_offset())
+            .with_slot(ffi::Py_tp_base, T::BaseType::type_object_raw(py))
+            .with_slot(ffi::Py_tp_dealloc, tp_dealloc::<T> as *mut c_void)
+            .set_is_basetype(T::IS_BASETYPE)
+            .set_is_mapping(T::IS_MAPPING)
+            .with_class_items(T::items_iter())
+            .build(py, T::NAME, T::MODULE, std::mem::size_of::<T::Layout>())
     } {
         Ok(type_object) => type_object,
         Err(e) => type_object_creation_failed(py, e, T::NAME),
     }
 }
 
-#[allow(clippy::too_many_arguments)]
-unsafe fn create_type_object_impl(
-    py: Python<'_>,
-    tp_doc: &str,
-    module_name: Option<&str>,
-    name: &str,
-    base_type_object: *mut ffi::PyTypeObject,
-    basicsize: usize,
-    tp_dealloc: ffi::destructor,
-    dict_offset: Option<ffi::Py_ssize_t>,
-    weaklist_offset: Option<ffi::Py_ssize_t>,
-    get_items_iter: fn() -> PyClassItemsIter,
-    is_basetype: bool,
+#[derive(Default)]
+struct PyTypeBuilder {
+    slots: Vec<ffi::PyType_Slot>,
+    method_defs: Vec<ffi::PyMethodDef>,
+    property_defs_map: HashMap<&'static str, ffi::PyGetSetDef>,
+    /// Used to patch the type objects for the things there's no
+    /// PyType_FromSpec API for... there's no reason this should work,
+    /// except for that it does and we have tests.
+    cleanup: Vec<Box<dyn Fn(&PyTypeBuilder, *mut ffi::PyTypeObject)>>,
     is_mapping: bool,
-) -> PyResult<*mut ffi::PyTypeObject> {
-    let mut slots = Vec::new();
+    has_new: bool,
+    has_dealloc: bool,
+    has_getitem: bool,
+    has_setitem: bool,
+    has_traverse: bool,
+    has_clear: bool,
+    has_dict: bool,
```

Comment:
When you have a lot of booleans it's often better to use enums. Something like:

```rust
enum Traverse{
    YES,
    NO,
}
```

Reply:
Given these are internal state of the builder, I think it's fine. Enums definitely improve readability in function calls, don't think that's applicable here though?

Reply:
Yeah I think it's fine either way.

Reply:
Yeah, if these were mutually exclusive in any way I'd make an enum variant for them, but as it is we'd just need to make 7 enum variants to represent true/false, and isn't worth the added effort.